### PR TITLE
Logging system

### PR DESCRIPTION
**Pull request by saveman71**
Still WIP. No need to review as of now.

### CODE REVIEW

File: include/lug/System/Logger.hpp [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+#include <lug/System/Logger/Message.hpp>
+#include <lug/System/Utils.hpp>
+
+namespace lug {
+namespace System {
+
+class LUG_SYSTEM_API Logger {
+public:
+    Logger(const std::string& loggerName);
+    virtual ~Logger() {}
```
Comment (Quentin01) [inline include/lug/System/Logger.hpp]: ` = default` and do you really want to inherit from `Logger`? If not, just remove all virtual functions.

Comment (Quentin01) [inline include/lug/System/Logger.hpp]: Define copy / move constructors / operators (can be `= delete`)

File: include/lug/System/Logger.hpp [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+    virtual void addHandler(Handler* handler) {
```
Comment (Quentin01) [inline include/lug/System/Logger.hpp]: If you want to define functions in the header, define them in a `.inl` file

File: include/lug/System/Logger.hpp [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        _handlers.insert(handler);
+    }
+
+    virtual void addHandler(const std::string& name) {
+        _handlers.insert(LoggingFacility::getHandler(name));
+    }
+
+    virtual void defaultErrHandler(const std::string& msg) {
```
Comment (Quentin01) [inline include/lug/System/Logger.hpp]: Maybe take an exception in parameter and not a `std::string`? Is it okay for an user to use it?

File: include/lug/System/Logger/Common.hpp [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    using filename_t = std::string;
+#endif
+
+namespace Channel {
+typedef enum {
+    LUG_LOG_CHANNELS(LUG_LOG_ENUM)
+} enumChannel;
+}
+
+namespace Level {
```
Comment (Quentin01) [inline include/lug/System/Logger/Common.hpp]: What? Just use an `enum class Level` no?

File: include/lug/System/Logger.hpp [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+        catch (const std::exception& ex) {
+            defaultErrHandler(ex.what());
+        }
+        catch (...) {
+            defaultErrHandler("Unknown exception");
+        }
+    }
+
+    template<typename... Args, typename T>
+    inline void log(Level::enumLevel lvl, const T& fmt, const Args &... args) {
```
Comment (Quentin01) [inline include/lug/System/Logger.hpp]: Use universal reference and forward them to `write`

File: include/lug/System/Logger.hpp [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+        catch (const std::exception& ex) {
+            defaultErrHandler(ex.what());
+        }
+        catch (...) {
+            defaultErrHandler("Unknown exception");
+        }
+    }
+
+    template<typename T, typename... Args>
+    inline void debug(const T& fmt, const Args&... args) {
```
Comment (Quentin01) [inline include/lug/System/Logger.hpp]: Use universal reference and forward them

File: include/lug/System/Logger.hpp [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+        log(Level::Debug, fmt, args...);
+    }
+
+    template<typename T, typename... Args>
+    inline void info(const T& fmt, const Args&... args) {
```
Comment (Quentin01) [inline include/lug/System/Logger.hpp]: Use universal reference and forward them

File: include/lug/System/Logger.hpp [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+        log(Level::Info, fmt, args...);
+    }
+
+    template<typename T, typename... Args>
+    inline void warn(const T& fmt, const Args&... args) {
```
Comment (Quentin01) [inline include/lug/System/Logger.hpp]: Use universal reference and forward them

File: include/lug/System/Logger.hpp [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+        log(Level::Warning, fmt, args...);
+    }
+
+    template<typename T, typename... Args>
+    inline void error(const T& fmt, const Args&... args) {
```
Comment (Quentin01) [inline include/lug/System/Logger.hpp]: Use universal reference and forward them

File: include/lug/System/Logger.hpp [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+        log(Level::Error, fmt, args...);
+    }
+
+    template<typename T, typename... Args>
+    inline void fatal(const T& fmt, const Args&... args) {
```
Comment (Quentin01) [inline include/lug/System/Logger.hpp]: Use universal reference and forward them

File: include/lug/System/Logger.hpp [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+        log(Level::Fatal, fmt, args...);
+    }
+
+    template<typename T, typename... Args>
+    inline void assrt(const T& fmt, const Args&... args) {
```
Comment (Quentin01) [inline include/lug/System/Logger.hpp]: Use universal reference and forward them

Comment (Quentin01) [inline include/lug/System/Logger.hpp]: Maybe `assert`?

File: include/lug/System/Logger/Common.hpp [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+#define LUG_LOG_MAP_PAIR_C(CHANNEL) { Logger::Channel::CHANNEL, #CHANNEL }
+#define LUG_LOG_MAP_PAIR_T(CHANNEL) { Level::CHANNEL, #CHANNEL }
+
+#if defined(LUG_SYSTEM_WINDOWS)
+    using filename_t = std::wstring;
+#else
+    using filename_t = std::string;
+#endif
+
+namespace Channel {
```
Comment (Quentin01) [inline include/lug/System/Logger/Common.hpp]: What? Just use an `enum class Channel` no?

File: include/lug/System/Logger/FileHandler.hpp [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+#include <fstream>
+#include <lug/System/Logger/Handler.hpp>
+#include <lug/System/Exception.hpp>
+
+namespace lug {
+namespace System {
+
+class FileHandler : public Handler {
+public:
+    FileHandler(const std::string& name, const filename_t& filename, bool truncate) : Handler(name) {
```
Comment (Quentin01) [inline include/lug/System/Logger/FileHandler.hpp]: Don't put the content of methods in the header, or use a file `.inl` for that

File: include/lug/System/Logger/FileHandler.hpp [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+        _ofs.open(filename, std::ofstream::out);
+        if (!_ofs.good()) {
+            // TODO: add file to lug except when it handles variadic args
+            LUG_EXCEPT(FileNotFoundException, "Failed to open file");
+        }
+        // TODO: use this parameter
+        (void)truncate;
+    }
+    virtual ~FileHandler() {
```
Comment (Quentin01) [inline include/lug/System/Logger/FileHandler.hpp]: Do you really want to inherit from that class? If not remove all the virtual functions.

File: include/lug/System/Logger/FileHandler.hpp [hunk 15]
```diff
@@ -0,0 +1,38 @@
+#pragma once
+
+#include <fstream>
+#include <lug/System/Logger/Handler.hpp>
+#include <lug/System/Exception.hpp>
+
+namespace lug {
+namespace System {
+
+class FileHandler : public Handler {
```
Comment (Quentin01) [inline include/lug/System/Logger/FileHandler.hpp]: Don't you want to export it?

File: include/lug/System/Logger/Formatter.hpp [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+    FlagHandlerPointer basic = nullptr;
+    std::shared_ptr<Formattable> advanced = nullptr;
+};
+
+} // namespace priv
+
+class LUG_SYSTEM_API Formatter {
+public:
+    Formatter(const std::string& pattern);
+    virtual ~Formatter () {};
```
Comment (Quentin01) [inline include/lug/System/Logger/Formatter.hpp]: `= default`

File: include/lug/System/Logger/Formatter.hpp [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+    std::string handleFlagY(const struct tm& now); // Year 4 digits
+    std::string handleFlagm(const struct tm& now); // Month from 1 to 12
+    std::string handleFlagd(const struct tm& now); // Day from 1 to 31
+    std::string handleFlagH(const struct tm& now); // Hour from 0 to 24
+    std::string handleFlagM(const struct tm& now); // Minute from 0 to 59
+    std::string handleFlagS(const struct tm& now); // Sec from 0 to 61
+
+    std::vector<priv::Token> _formatChain;
+};
+
```
Comment (Quentin01) [inline include/lug/System/Logger/Formatter.hpp:75]: Remove the double vertical space

File: include/lug/System/Logger/Handler.hpp [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+#include <lug/System/Logger/Message.hpp>
+#include <lug/System/Logger/Formatter.hpp>
+#include <lug/System/Logger/LoggingFacility.hpp>
+
+namespace lug {
+namespace System {
+
+class LUG_SYSTEM_API Handler {
+public:
+    Handler(const std::string& name);
```
Comment (Quentin01) [inline include/lug/System/Logger/Handler.hpp:16]: Implement copy / move constructors / operators

File: include/lug/System/Logger/LogCatHandler.hpp [hunk 19]
```diff
@@ -0,0 +1,40 @@
...
+
+#include <android/log.h>
+#include <lug/System/Logger/Handler.hpp>
+
+namespace lug {
+namespace System {
+
+class LogCatHandler : public Handler {
+public:
+    LogCatHandler(const std::string& name) : Handler(name) {
```
Comment (Quentin01) [inline include/lug/System/Logger/LogCatHandler.hpp]: Indent correctly the brackets

Comment (Quentin01) [inline include/lug/System/Logger/LogCatHandler.hpp]: Don't put the content of methods in the header, or use a file `.inl` for that

File: include/lug/System/Logger/LogCatHandler.hpp [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+    }
+    virtual ~LogCatHandler() {
```
Comment (Quentin01) [inline include/lug/System/Logger/LogCatHandler.hpp]: `= default` and do you really want to inherit from that class?

File: include/lug/System/Logger/Message.hpp [hunk 21]
```diff
@@ -0,0 +1,23 @@
...
+
+#include <lug/System/Logger/Common.hpp>
+
+namespace lug {
+namespace System {
+namespace priv {
+
+struct Message {
+    Message() = default;
+    Message(const std::string& _loggerName, Level::enumLevel _level): loggerName(_loggerName), level(_level) {
```
Comment (Quentin01) [inline include/lug/System/Logger/Message.hpp]: Indent correctly the brackets

File: include/lug/System/Logger/OstreamHandler.hpp [hunk 22]
```diff
@@ -0,0 +1,38 @@
...
+
+#include <iostream>
+#include <lug/System/Logger/Handler.hpp>
+
+namespace lug {
+namespace System {
+
+class OstreamHandler : public Handler {
+public:
+    OstreamHandler(const std::string& name, std::ostream& out) : Handler(name), _os(out) {
```
Comment (Quentin01) [inline include/lug/System/Logger/OstreamHandler.hpp]: Indent correctly the brackets

File: include/lug/System/Logger/OstreamHandler.hpp [hunk 23]
```diff
@@ -0,0 +1,38 @@
...
+    }
+    virtual ~OstreamHandler() {
```
Comment (Quentin01) [inline include/lug/System/Logger/OstreamHandler.hpp]: Indent correctly the brackets

File: include/lug/System/Logger/FileHandler.hpp (hunk 15 above)
Comment (Quentin01) [inline include/lug/System/Logger/FileHandler.hpp]: Why aren't you inheriting from `OstreamHandler` ?

File: include/lug/System/Logger/Message.hpp [hunk 24]
```diff
@@ -0,0 +1,23 @@
+#pragma once
+
+#include <lug/System/Logger/Common.hpp>
+
+namespace lug {
+namespace System {
+namespace priv {
+
+struct Message {
```
Comment (Quentin01) [inline include/lug/System/Logger/Message.hpp]: Don't you want to export it?

File: include/lug/System/Logger/LogCatHandler.hpp [hunk 25]
```diff
@@ -0,0 +1,40 @@
+#pragma once
+
+#include <android/log.h>
+#include <lug/System/Logger/Handler.hpp>
+
+namespace lug {
+namespace System {
+
+class LogCatHandler : public Handler {
```
Comment (Quentin01) [inline include/lug/System/Logger/LogCatHandler.hpp]: Don't you want to export it?

File: samples/base/main.cpp [hunk 26]
```diff
@@ -1,13 +1,21 @@
 #include <lug/Config.hpp>
 #include <lug/System/Exception.hpp>
+#include <lug/System/Logger.hpp>
 #include <iostream>
 
-int main() {
-    try {
+int main(int, const char*[]) {
```
Comment (Quentin01) [inline samples/base/main.cpp:6]: Remove all the useless comments and remove the `const` (It can be a problem in the future)

Comment (Quentin01) [inline samples/base/main.cpp:6]: Maybe log the exception to demonstrate the logger?

File: samples/window/main.cpp [hunk 27]
```diff
@@ -1,20 +1,32 @@
 #include <lug/Window/Window.hpp>
+#include <lug/System/Logger.hpp>
+#include <lug/System/Logger/LogCatHandler.hpp>
```
Comment (Quentin01) [inline samples/window/main.cpp]: Be sure to be on Android for that!

File: src/lug/System/CMakeLists.txt [hunk 28]
```diff
@@ -16,6 +20,15 @@ set(INC
...
     ${INCROOT}/Export.hpp
+    ${INCROOT}/Logger.hpp
+    ${INCROOT}/Logger/Common.hpp
+    ${INCROOT}/Logger/FileHandler.hpp
+    ${INCROOT}/Logger/Formatter.hpp
+    ${INCROOT}/Logger/Handler.hpp
+    ${INCROOT}/Logger/LoggingFacility.hpp
+    ${INCROOT}/Logger/Message.hpp
+    ${INCROOT}/Logger/OstreamHandler.hpp
+    ${INCROOT}/Logger/LogCatHandler.hpp
```
Comment (Quentin01) [inline src/lug/System/CMakeLists.txt]: Only for Android?

File: src/lug/System/Logger/Formatter.cpp [hunk 29]
```diff
@@ -0,0 +1,157 @@
...
+    struct tm timeInfo;
+
+    msg.formatted.clear();
+
+    #if defined(LUG_SYSTEM_WINDOWS)
+    // Use windows secure versions of localtime
+        localtime_s(&timeInfo, &tt);
+    #else
+    // Use linux secure versions of localtime
+    // TODO: test on android
```
Comment (Quentin01) [inline src/lug/System/Logger/Formatter.cpp]: So, tested or not?

File: src/lug/System/Logger/Formatter.cpp [hunk 30]
```diff
@@ -0,0 +1,157 @@
...
+    // It could not work (what is the localtime secure version for android ?)
+        localtime_r(&tt, &timeInfo);
+    #endif
+
+    for (priv::Token elem : _formatChain) {
+        if (!elem.basic && !elem.advanced) {
+            msg.formatted << msg.raw.str();
+        }
+        else if (elem.basic) {
```
Comment (Quentin01) [inline src/lug/System/Logger/Formatter.cpp]: After the closing bracket

File: src/lug/System/Logger/Formatter.cpp [hunk 31]
```diff
@@ -0,0 +1,157 @@
...
+            msg.formatted << (this->*elem.basic)(timeInfo);
+        }
+        else if (elem.advanced) {
```
Comment (Quentin01) [inline src/lug/System/Logger/Formatter.cpp]: After the closing bracket

File: src/lug/System/Logger/Handler.cpp [hunk 32]
```diff
@@ -0,0 +1,39 @@
...
+namespace lug {
+namespace System {
+
+Handler::Handler(const std::string& name):
+    _name(name),
+    _formatter(std::make_unique<Formatter>("%v\n")),
+    _level(Level::Debug)
+{}
+
+Handler::~Handler() {}
```
Comment (Quentin01) [inline src/lug/System/Logger/Handler.cpp]: `= default`

File: src/lug/System/Logger/Logger.cpp [hunk 33]
```diff
@@ -0,0 +1,30 @@
+#include <lug/System/Logger.hpp>
+
+namespace lug {
+namespace System {
+
+Logger Logger::logger("internal");
+
+Logger::Logger(const std::string& loggerName) : _name(loggerName) {}
+
+const std::string& Logger::getName() const {
```
Comment (Quentin01) [inline src/lug/System/Logger/Logger.cpp:27]: Inline getters maybe

File: test/System/Logger/MockHandler.hpp [hunk 34]
```diff
@@ -0,0 +1,21 @@
...
+
+#include <lug/System/Logger/Handler.hpp>
+#include <gmock/gmock.h>
+
+namespace lug {
+namespace System {
+
+class MockHandler : public Handler {
+public:
+    MockHandler(const std::string& name) : Handler(name) {
```
Comment (Quentin01) [inline test/System/Logger/MockHandler.hpp]: Indent your brackets correctly

File: test/System/Logger/MockHandler.hpp [hunk 35]
```diff
@@ -0,0 +1,21 @@
...
+    }
+    virtual ~MockHandler() {
```
Comment (Quentin01) [inline test/System/Logger/MockHandler.hpp]: Do you really want to inherit from that class?

**Review by Quentin01 — CHANGES_REQUESTED** [review]

File: include/lug/System/Logger/Message.hpp [hunk 36]
```diff
@@ -0,0 +1,23 @@
...
+
+    const std::string loggerName;
+    Level::enumLevel level;
+    fmt::MemoryWriter raw;
+    fmt::MemoryWriter formatted;
+};
+
+}
+}
+}
```
Comment (Sumenia) [inline include/lug/System/Logger/Message.hpp]: specify the closing namespace in comment

File: include/lug/System/Logger/Formatter.hpp (hunk 17 above)
Comment (saveman71) [reply include/lug/System/Logger/Formatter.hpp:75]: Done

File: include/lug/System/Logger/Handler.hpp (hunk 18 above)
Comment (saveman71) [reply include/lug/System/Logger/Handler.hpp:16]: Done

**Review by Quentin01 — APPROVED** [review]

**Pull request merged**